File: spark/python_code/ingestionframework-master/src/publish/write_file.py
# -*- coding: utf-8 -*-
"""
Created on 04/12/2020 2:01 PM

@author : Nirav Langaliya
"""
from typing import List

from pyspark.sql import DataFrame

# from src.config.config_objects import ColumnNameType
# from src.dependencies.dataframe_utils import dataframe_add_columns
# from src.dependencies.schema_utilties import get_options

from config.config_objects import ColumnNameTypeLength
from dependencies.dataframe_utils import dataframe_add_columns
from dependencies.schema_utilties import get_options
from dependencies.columnTranformation import snakeCaseHeaderFun

def file_writer(
        dataframe:DataFrame,
        format,
        path,
        mode,
        header,
        snakeCaseHeader,
        delimiter,
        num_partitions,
        partition_columns,
        addColumns: List[ColumnNameTypeLength],
        selectCols
):

    if selectCols:
        print(selectCols)
        dataframe = dataframe.select(selectCols)

    # Add columns
    if addColumns:
        dataframe = dataframe_add_columns(dataframe,addColumns)

    if snakeCaseHeader:
        dataframe = snakeCaseHeaderFun(dataframe)

    # Repartition dataframe
    if num_partitions > 0:
        dataframe = dataframe.repartition(num_partitions) #.write.mode(mode).partitionBy(partition_columns)

    # Partition By column list
    if partition_columns:
        writer = dataframe.write.mode(mode).partitionBy(partition_columns)
    else:
        writer = dataframe.write.mode(mode)

    options = get_options(
        header = header, delimiter = delimiter
    )

    for option_name in options:
        writer = writer.option(option_name, options[option_name])

    if format.lower() == 'csv':
        writer.csv(path)
    elif format.lower() == "json":
        writer.json(path)
    elif format.lower() == "parquet":
        writer.parquet(path)
    elif format.lower() == "avro":
        writer.format("avro").save(path)
    elif format.lower() == "delta":
        writer.format("delta").save(path)
    elif format.lower() == "xml":
        writer.xml(path)
    elif format.lower() == "text":
        writer.text(path)
        pass
    else:
        raise Exception(" Unknown data write format! ")

File: spark/javaingestionframework-master/src/main/java/com/nisum/dataingestionframework/config/WriteFile.java
package com.nisum.dataingestionframework.config;

import com.nisum.dataingestionframework.dependencies.ColumnTranformation;
import com.nisum.dataingestionframework.dependencies.DataFrameUtils;
import com.nisum.dataingestionframework.dependencies.DateTimeUtilities;
import com.nisum.dataingestionframework.dependencies.SchemaUtilities;
import com.nisum.dataingestionframework.exception.ResourceNotFoundException;
import com.nisum.dataingestionframework.model.FileOutput;
import com.nisum.dataingestionframework.util.Constants;
import org.apache.spark.sql.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Serializable;
import java.util.*;

/* @author Hari Chandra Prasad Nimmagadda */

public class WriteFile implements Serializable {

    private static final Logger logger = LoggerFactory.getLogger(WriteFile.class);

    DataFrameUtils dataFrameUtils = new DataFrameUtils();
    SchemaUtilities schemaUtilities = new SchemaUtilities();
    ColumnTranformation columnTranformation = new ColumnTranformation();
    DateTimeUtilities dateTimeUtilities = new DateTimeUtilities();
    String res = new String();

    public String fileWriter(String dataFrameName, FileOutput fileOutput, Map<String, Dataset<Row>> datasetModelMap) {

        logger.info(" mapModel file writer  .... " + datasetModelMap.toString());

        datasetModelMap.entrySet().stream().forEach(datasetModel -> {

            if (datasetModel.getKey().equalsIgnoreCase(dataFrameName)) {

                Dataset<Row> dataFrame = datasetModel.getValue();

                if (fileOutput.getSelectCol().size() > 0) {

                    Column[] selectedColumns = new Column[fileOutput.getSelectCol().size()];
                    for (int i = 0; i < selectedColumns.length; i++) {
                        selectedColumns[i] = new Column(fileOutput.getSelectCol().get(i));
                    }

                    dataFrame = dataFrame.select(selectedColumns);

                }

                //  Add Columns
                if (fileOutput.getAddColumns().size() > 0) {
                    dataFrame = dataFrameUtils.dataframeAddColumns(dataFrame, fileOutput.getAddColumns());
                }

                //  SnakeCaseHeader
                if (fileOutput.isSnakeCaseHeader())
                    dataFrame = columnTranformation.snakeCaseHeaderFun(dataFrame);

                // Repartition dataframe
                if (fileOutput.getNumPartitions() > 0)
                    dataFrame = dataFrame.repartition(fileOutput.getNumPartitions());

                //  Partition By column list

             if (fileOutput.getPartitionBy()!=null) {
                 List<String> list = fileOutput.getPartitionBy();
                 String delim = ",";

                  res = String.join(delim, list);
                }

                boolean header = fileOutput.isHeader();
                String delimiter = fileOutput.getDelimiter();

                Map optionsMap = schemaUtilities.getOptions(header, delimiter);

                String fileFormat = fileOutput.getFormat().toLowerCase();
                String path = fileOutput.getFileOutPath();

                logger.info("Before Execution of Write File " + dateTimeUtilities.getCurrentDateTime());

                dataFrame.show();
                if (fileFormat.toLowerCase().equalsIgnoreCase(Constants.CSV)) {
                    dataFrame.write().options(optionsMap).partitionBy(res).csv(path);
                } else if (fileFormat.toLowerCase().equalsIgnoreCase(Constants.JSON))
                    dataFrame.write().json(path);
                else if (fileFormat.toLowerCase().equalsIgnoreCase(Constants.PARQUET))
                    dataFrame.write().parquet(path);
                else if (fileFormat.toLowerCase().equalsIgnoreCase(Constants.AVRO))
                    dataFrame.write().format(Constants.AVRO).save(path);
                else if (fileFormat.toLowerCase().equalsIgnoreCase(Constants.DELTA))
                    dataFrame.write().format(Constants.DELTA).save(path);
                else if (fileFormat.toLowerCase().equalsIgnoreCase(Constants.XML))
                    dataFrame.write().format("com.databricks.spark.xml").save(path);
                else if (fileFormat.toLowerCase().equalsIgnoreCase(Constants.TEXT))
                    dataFrame.write().text(path);
                else if (fileFormat.toLowerCase().equalsIgnoreCase(Constants.FIXED))
                    dataFrame.write().options(optionsMap).csv(path);
                else
                    try {
                        throw new ResourceNotFoundException(" Unknown data write format! ");
                    } catch (ResourceNotFoundException e) {
                        e.printStackTrace();
                    }
            }



        });

            logger.info("After Execution of Write File " + dateTimeUtilities.getCurrentDateTime());


        return "Failed to write data";
    }

}
